<a href="https://colab.research.google.com/github/ivazquez-07/python_data_science_excersises/blob/main/Curso-de-Estadistica-con-Python-Aula-1/Curso_de_Estad%C3%ADstica_con_Python_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***
# <font color=#56B4E9 size=10>CURSO DE ESTADÍSTICA</font>
***

<font color=#56B4E9 size = 5>*Preparando el ambiente*</font>

# <font color=#56B4E9>1 CONOCIENDO LOS DATOS</font>
***

## <font color=#56B4E9>1.1 Dataset del proyecto</font>
***

### Muestra de domicilios Colombia - 2018

Las investigaciones por muestras de domicilios realizadas anualmente, buscan encontrar característcias generales de la población, de educación, trabajo, rendimiento y otras, de acuerdo con las necesidades de información del país, tales como las características de migración, fertilidad, casamientos, salud, nutrición, entre otros temas. Estas muestras al pasar de los años consistuyen una herramienta importante para la formulación, validación y evaluación de políticas dirigidas al desarrollo socioeconómico y la mejora de las condiciones de vida en Colombia.


### Datos

Los datos fueron creados  de manera didáctica para este curso.

### Variables utilizadas

> ### Ingreso
> ***

Ingresos mensuales (en miles de pesos) del trabajo principal para personas de 10 años o más.

> ### Edad
> ***

Edad del entrevistado en la fecha de referencia en años.

> ### Altura
> ***

Altura del entrevistado em metros.

> ### Ciudad
> ***

Código de referência a 27 ciudades analizadas.


> ### Sexo
> ***

|Código|Descripción|
|---|---|
|0|Masculino|
|1|Femenino|

> ### Años de Estudio
> ***

|Código|Descripción|
|---|---|
|1|Sin estudios y menos de 1 año|
|2|1 año|
|3|2 años|
|4|3 años|
|5|4 años|
|6|5 años|
|7|6 años|
|8|7 años|
|9|8 años|
|10|9 años|
|11|10 años|
|12|11 años|
|13|12 años|
|14|13 años|
|15|14 años|
|16|15 años o más|
|17|No se sabe|
||No aplica|

> ### Color
> ***

|Código|Descripción|
|---|---|
|0|Indio|
|2|Blanco|
|4|Negro|
|6|Amarillo|
|8|Moreno|
|9|Sin declarar|

> ### Tratamiento a los datos
***

> Algunos de los tratamientos de datos más frecuentes son:
> 1. Eliminar las observaciones (líneas) con entradas de datos inválidos;
> 2. Eliminar observaciones donde hay datos perdidos (missing data);
> 3. Filtros própios de la investigación, por ejemplo: considerar solo las encuestas realizadas a la cabeza de familia (responsable por el domicilio).

### Importando pandas y leyendo el dataset del projecto

https://pandas.pydata.org/

In [11]:
import pandas as pd

In [12]:
datos = pd.read_csv("/content/datos.csv")

In [13]:
datos.sample(10)

Ciudad  Sexo  Edad  Color  Años de Estudio  Ingreso    Altura
13940      22     0    47      8                2      800  1.690110
39373      31     0    54      8               12     1980  1.636566
24437      27     1    30      2               12      809  1.666931
59068      41     0    40      2               12     3500  1.594617
27440      29     0    41      4               12     1200  1.670998
14988      23     0    30      8               12     1800  1.797236
75885      53     0    47      8               12     1800  1.729711
2305       12     1    43      8                1      788  1.794999
26100      28     0    43      8                1      290  1.670435
67921      43     0    40      2                6     1800  1.747748

## <font color=#56B4E9>1.2 Tipos de datos</font>
***

### Clasificación de una variable
<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img001.png' width='70%'>

### Variables cualitativas ordinales

► Variables que pueden ser ordenadas o que responen algun tipo de jerarquía

In [14]:
sorted(datos['Años de Estudio'].unique())

[np.int64(1),
 np.int64(2),
 np.int64(3),
 np.int64(4),
 np.int64(5),
 np.int64(6),
 np.int64(7),
 np.int64(8),
 np.int64(9),
 np.int64(10),
 np.int64(11),
 np.int64(12),
 np.int64(13),
 np.int64(14),
 np.int64(15),
 np.int64(16),
 np.int64(17)]

### Variables cualitativas nominales

► Variables que no pueden ser ordenadas ni responden alguna jerarquía.

In [15]:
sorted(datos['Sexo'].unique())

[np.int64(0), np.int64(1)]

In [16]:
sorted(datos['Color'].unique())

[np.int64(0), np.int64(2), np.int64(4), np.int64(6), np.int64(8)]

In [17]:
sorted(datos['Ciudad'].unique())

[np.int64(11),
 np.int64(12),
 np.int64(13),
 np.int64(14),
 np.int64(15),
 np.int64(16),
 np.int64(17),
 np.int64(21),
 np.int64(22),
 np.int64(23),
 np.int64(24),
 np.int64(25),
 np.int64(26),
 np.int64(27),
 np.int64(28),
 np.int64(29),
 np.int64(31),
 np.int64(32),
 np.int64(33),
 np.int64(35),
 np.int64(41),
 np.int64(42),
 np.int64(43),
 np.int64(50),
 np.int64(51),
 np.int64(52),
 np.int64(53)]

### Variables cuantitativas discretas

► Variables que representan un conteo donde los valores posibles forman un conjunto finito o numerable.


In [18]:
datos.Edad.max()

99

In [19]:
print(f"La edad mínima es {datos.Edad.min()} y la edad máxima es {datos.Edad.max()}")

La edad mínima es 13 y la edad máxima es 99


#### <font color='red'>Observación</font>
***
> La variable edad puede ser clasificada de tres formas diferentes:
> 1. <b>CUANTITATIVA DISCRETA</b> - cuando representa años completos (números enteros);
> 2. <b>CUANTITATIVA CONTINUA</b> - cuando representa la edad exacta, siendo representada por fracciones de años;
> 3. <b>CUALITATIVA ORDINAL</b> - cuando representa  intervalos de edad.

### Variables cuantitativas continuas

► Variables que representan un conteo o medición que asume valores en escalas continuas (números reales).

In [20]:
print(f"La Altura mínima es {datos['Altura'].min()} y la altura máxima es {datos['Altura'].max()}")

La Altura mínima es 1.339244614 y la altura máxima es 2.028496765


# <font color=#56B4E9>2 DISTRIBUCIÓN DE FRECUENCIAS</font>
***

El primer paso en un análisis es conocer el comportamiento de las variables involucradas en el estudio. Utilizando técnicas estadísticas como el análisis de las <b> DISTRIBUCIONES DE FRECUENCIAS </b> e <b> HISTOGRAMAS </b> podemos evaluar mejor cómo se distribuyen los fenómenos del estudio.



## <font color=#56B4E9>2.1 Distribución de frecuencias para variables cualitativas</font>
***

### Método 1

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.Series.value_counts.html

In [21]:
datos.Sexo.value_counts()

Sexo
0    53250
1    23590
Name: count, dtype: int64

In [22]:
porcentaje_sexo = (datos.Sexo.value_counts(normalize=True)*100).round(4)
porcentaje_sexo


Sexo
0    69.2998
1    30.7002
Name: proportion, dtype: float64

In [23]:
dist_frec_cualitativa = pd.DataFrame({'Sexo_id': datos.Sexo.value_counts().index,
              'Cantidad': datos.Sexo.value_counts().values,
              'Porcentaje (%)': porcentaje_sexo})

dist_frec_cualitativa.rename(index=  {0:'Masculino', 1:'Femenino'}, inplace=True)

dist_frec_cualitativa

Sexo_id  Cantidad  Porcentaje (%)
Sexo                                        
Masculino        0     53250         69.2998
Femenino         1     23590         30.7002

### Método 2

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.crosstab.html

In [24]:
sexo = {0: 'Masculino',
        1: 'Femenino'}

color = {0: 'Indigena',
         2: 'Blanco',
         4: 'Negro',
         6: 'Amarillo',
         8: 'Pardo',
         9: 'Sin declarar'}

In [25]:
frecuencia = pd.crosstab(datos.Sexo, datos.Color)

In [26]:
frecuencia.rename(index= sexo, inplace= True)

In [27]:
frecuencia.rename(columns= color, inplace= True)

In [28]:
frecuencia

Color      Indigena  Blanco  Negro  Amarillo  Pardo
Sexo                                               
Masculino       256   22194   5502       235  25063
Femenino        101    9621   2889       117  10862

In [29]:
porc_freq = (pd.crosstab(datos.Sexo, datos.Color, normalize= True)*100).round(2).astype(str) + ' %'
porc_freq.rename(index= sexo, inplace= True)
porc_freq.rename(columns= color, inplace= True)
porc_freq

Color     Indigena   Blanco   Negro Amarillo    Pardo
Sexo                                                 
Masculino   0.33 %  28.88 %  7.16 %   0.31 %  32.62 %
Femenino    0.13 %  12.52 %  3.76 %   0.15 %  14.14 %

In [30]:
porcentaje = pd.crosstab(datos.Sexo,
                         datos.Color,
                         aggfunc = 'mean',
                         values = datos.Ingreso).round(2).astype(str)
porcentaje.rename(index = sexo,inplace = True)
porcentaje.rename(columns = color,inplace = True)
'$ ' + porcentaje

Color       Indigena     Blanco      Negro   Amarillo      Pardo
Sexo                                                            
Masculino  $ 1081.71  $ 2925.74  $ 1603.86  $ 4758.25  $ 1659.58
Femenino   $ 2464.39  $ 2109.87   $ 1134.6  $ 3027.34  $ 1176.76

## <font color=#56B4E9>2.2 Distribución de frecuencia para variables cuantitativas (clases personalizadas)</font>
***

### Paso 1 - Especifique los límites de cada clase

Utilice la siguiente clasificación:

<b>A</b> ► Más de 20 SM

<b>B</b> ► De 10 a 20 SM

<b>C</b> ► De 4 a 10 SM

<b>D</b> ► De 2 a 4 SM

<b>E</b> ► Hasta 2 SM

donde <b>SM</b> es el valor del salario mínimo en ese momento. En nuestro caso <b>788 mil pesos colombianos</b> (2018):

<b>A</b> ► Más de 15.760

<b>B</b> ► De 7.880 a 15.760

<b>C</b> ► De 3.152 a 7.880

<b>D</b> ► De 1.576 a 3.152

<b>E</b> ► Hasta 1.576


In [31]:
datos.Ingreso.min()

0

In [32]:
datos.Ingreso.max()

200000

In [33]:
clases = [0, 1576, 3152, 7880, 15760, 200000]

In [34]:
label = ['E', 'D', 'C', 'B', 'A']

In [35]:
clases

[0, 1576, 3152, 7880, 15760, 200000]

In [36]:
label

['E', 'D', 'C', 'B', 'A']

### Paso 2 - Crear la tabla de frecuencias

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.cut.html

In [54]:
frecuencia = pd.value_counts(pd.cut(x = datos.Ingreso,
                                    bins = clases,
                                    labels = label,
                                    include_lowest = True))

/tmp/ipython-input-54-3662911492.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  frecuencia = pd.value_counts(pd.cut(x = datos.Ingreso,


In [55]:
frecuencia.sort_index(ascending = False, inplace = True)

frecuencia

Ingreso
A      608
B     2178
C     7599
D    16700
E    49755
Name: count, dtype: int64

In [56]:
porcentaje = (pd.value_counts(pd.cut(x = datos.Ingreso,
                                    bins = clases,
                                    labels = label,
                                    include_lowest = True),
                             normalize=True)*100).round(2)
porcentaje.sort_values().astype(str) + ' %'


/tmp/ipython-input-56-3058546122.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  porcentaje = (pd.value_counts(pd.cut(x = datos.Ingreso,


Ingreso
A     0.79 %
B     2.83 %
C     9.89 %
D    21.73 %
E    64.75 %
Name: proportion, dtype: object

In [57]:
dist_frec_cuantitativa_personalizada = pd.DataFrame({'Frecuencia': frecuencia,
                                                     'Porcentaje (%)': porcentaje})
dist_frec_cuantitativa_personalizada.sort_index(ascending = False, inplace = True)
dist_frec_cuantitativa_personalizada

Frecuencia  Porcentaje (%)
Ingreso                            
A               608            0.79
B              2178            2.83
C              7599            9.89
D             16700           21.73
E             49755           64.75

## <font color=#56B4E9>2.3 Distribución de frecuencia para variables cuantitativas (clases de amplitud fija)</font>
***

### Importando bibliotecas

http://www.numpy.org/

In [58]:
import numpy as np

### Paso 1 - Difiniendo el número de clases

#### Regla de Sturges

La regla de Sturges, propuesta por Herbert Sturges en 1926, es una regla práctica acerca del número de clases que deben considerar al elaborarse un histograma.

# $$k = 1 + \frac {10}{3}\log_{10}n$$


En esta expresión:

- $k$ es el número de clases.
- $n$ es el número total de observaciones de la muestra.
- $log_{10}$ es el logaritmo común de base 10.


El valor de k (número de clases) debe ser redondeado a un número entero.
Es común redondearlo con el siguiente criterio: si el entero del resultado previo al redondeo es "par" se redondea a la alta y si el entero es "impar" se redondea a la baja, resultando siempre un número impar de clases (lo que permite ver la acumulación cerca de la media en una distribución normal).

In [63]:
n = datos.shape[0]
n

76840

In [64]:
k = 1 + (10/3) * np.log10(n)

In [68]:
k

np.float64(17.285291187298853)

### Paso 2 - Crear la tabla de frecuencias

In [70]:
frecuencia_sturges = pd.value_counts(
    pd.cut(x = datos.Ingreso,
           bins = int(k),
           include_lowest = True))

frecuencia_sturges

/tmp/ipython-input-70-2098586931.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  frecuencia_sturges = pd.value_counts(


Ingreso
(-200.001, 11764.706]       75594
(11764.706, 23529.412]       1022
(23529.412, 35294.118]        169
(35294.118, 47058.824]         19
(47058.824, 58823.529]         16
(94117.647, 105882.353]         6
(58823.529, 70588.235]          5
(70588.235, 82352.941]          4
(188235.294, 200000.0]          3
(82352.941, 94117.647]          1
(117647.059, 129411.765]        1
(129411.765, 141176.471]        0
(105882.353, 117647.059]        0
(141176.471, 152941.176]        0
(152941.176, 164705.882]        0
(164705.882, 176470.588]        0
(176470.588, 188235.294]        0
Name: count, dtype: int64

In [75]:
porcentaje_sturges = (pd.value_counts(pd.cut(x = datos.Ingreso,
                                    bins = int(k),
                                    include_lowest = True),
                             normalize=True)*100).round(4)
porcentaje_sturges.sort_values().astype(str) + ' %'


/tmp/ipython-input-75-2492759462.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  porcentaje_sturges = (pd.value_counts(pd.cut(x = datos.Ingreso,


Ingreso
(105882.353, 117647.059]        0.0 %
(141176.471, 152941.176]        0.0 %
(152941.176, 164705.882]        0.0 %
(164705.882, 176470.588]        0.0 %
(129411.765, 141176.471]        0.0 %
(176470.588, 188235.294]        0.0 %
(117647.059, 129411.765]     0.0013 %
(82352.941, 94117.647]       0.0013 %
(188235.294, 200000.0]       0.0039 %
(70588.235, 82352.941]       0.0052 %
(58823.529, 70588.235]       0.0065 %
(94117.647, 105882.353]      0.0078 %
(47058.824, 58823.529]       0.0208 %
(35294.118, 47058.824]       0.0247 %
(23529.412, 35294.118]       0.2199 %
(11764.706, 23529.412]         1.33 %
(-200.001, 11764.706]       98.3784 %
Name: proportion, dtype: object

In [76]:

dist_frec_cuant_personalizada = pd.DataFrame({'Frecuencia': frecuencia_sturges,
                                                     'Porcentaje (%)': porcentaje_sturges})
dist_frec_cuant_personalizada.sort_index(ascending = False, inplace = True)
dist_frec_cuant_personalizada

Frecuencia  Porcentaje (%)
Ingreso                                             
(188235.294, 200000.0]             3          0.0039
(176470.588, 188235.294]           0          0.0000
(164705.882, 176470.588]           0          0.0000
(152941.176, 164705.882]           0          0.0000
(141176.471, 152941.176]           0          0.0000
(129411.765, 141176.471]           0          0.0000
(117647.059, 129411.765]           1          0.0013
(105882.353, 117647.059]           0          0.0000
(94117.647, 105882.353]            6          0.0078
(82352.941, 94117.647]             1          0.0013
(70588.235, 82352.941]             4          0.0052
(58823.529, 70588.235]             5          0.0065
(47058.824, 58823.529]            16          0.0208
(35294.118, 47058.824]            19          0.0247
(23529.412, 35294.118]           169          0.2199
(11764.706, 23529.412]          1022          1.3300
(-200.001, 11764.706]          75594         98.3784

## <font color=#56B4E9>2.4 Histograma</font>
***

El <b>HISTOGRAMA</b> es la representación gráfica de una distribución de frecuencia. Es un gráfico formado por un conjunto de rectángulos colocados uno al lado del otro, donde el área de cada rectángulo es proporcional a la frecuencia de la clase que representa.

### Importando la biblioteca

https://seaborn.pydata.org/

# <font color=#56B4E9>3 MEDIDAS DE TENDENCIA CENTRAL</font>
***

## DataFrame de ejemplo

In [ ]:
df = pd.DataFrame(data = {'María': [8, 10, 4, 8, 6, 10, 8],
                          'Pedro': [10, 2, 0.5, 1, 3, 9.5, 10],
                          'Pablo': [7.5, 8, 7, 8, 8, 8.5, 7]},
                  index = ['Matemática',
                           'Portugués',
                           'Inglés',
                           'Geografía',
                           'Historia',
                           'Física',
                           'Química'])
df.rename_axis('Asignaturas', axis = 'columns', inplace = True)
df

## <font color=#56B4E9>3.1 Media aritmética</font>
***

Es representada por $\mu$ cuando se refiere a la población y por $\bar{X}$ cuando se refiere a la muestra

# $$\mu = \frac 1n\sum_{i=1}^{n}X_i$$

donde

$n$ = número de observaciones (registros)

$X_i$ = valor de la i-ésima observación (registro)

## <font color=#56B4E9>3.2 Mediana</font>
***

Para obtener la mediana de un conjunto de datos, debemos proceder de la siguiente manera:
1. Ordenar el conjunto de datos;
2. Identificar el número de observaciones (registros) del conjunto de datos ($n$);
3. Identicar el elemento del medio:

> Cuando $n$ sea impar, la posición del elemento del medio se obtendrá de la siguiente manera:


# $$Elemento_{Md} = \frac{n+1}{2}$$

> Cuando $n$ sea par, la posición del elemento mediano se obtendrá de la siguiente manera:


# $$Elemento_{Md} = \frac{n}2$$

4. Obtener la mediana:

> Cuando $n$ sea impar:


# $$Md = X_{Elemento_{Md}}$$

> Cuando $n$ sea par:


# $$Md = \frac{X_{Elemento_{Md}} + X_{Elemento_{Md}+1}}2$$
***

### Ejemplo 1 - n impar

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img002.png' width='40%' style="float:left">

### Ejemplo 2 - n par

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img003.png' width='50%' style="float:left">

### Obtener la mediana en nuestro conjunto de datos

## <font color=#56B4E9>3.3 Moda</font>
***

La moda se puede definir como el valor más frecuente de un conjunto de datos. La moda es ampliamente utilizada para datos cualitativos.

### Obteniendo la moda de nuestro dataset

## <font color=#56B4E9>3.4 Relación entre media, mediana e moda</font>
***

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img004.png' width='80%'>

### Evaluando la variable INGRESO

***

### Evaluando la variable ALTURA

***

### Evaluando la variable AÑOS DE ESTUDIO

# <font color=#56B4E9>4 MEDIDAS DE LOCALIZACIÓN</font>
***

## <font color=#56B4E9>4.1 Cuartiles, deciles y percentiles</font>
***

Hay una serie de medidas de posición similares en su diseño a la mediana, aunque no son medidas de tendencia central. Como se sabe, la mediana divide la distribución en dos partes iguales en términos del número de elementos en cada parte. Los cuartiles, por otro lado, permiten que la distribución se divida en cuatro partes iguales en términos del número de elementos en cada uno; deciles en diez partes y centiles en cien partes iguales.

## <font color=#56B4E9>4.2 Box-plot</font>
***

El *box plot* da una idea de la posición, dispersión, asimetría, colas y valores extremos. La posición central está dada por la mediana y la dispersión por $IIQ$ (distancia inter cuartil). Las posiciones relativas de $Q1$, $Mediana$ y $Q3$ dan una idea de la simetría de la distribución. Las longitudes de las colas están dadas por las líneas que van desde el rectángulo a los valores remotos y por los valores extremos.

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img005.png' width='65%'>

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img006.png' width='80%'>

# <font color=#56B4E9>5 MEDIDAS DE DISPERSIÓN</font>
***

Aunque las medidas de tendencia central y de localización proporcionan un resumen muy importante de los datos, pueden no ser suficientes para caracterizar diferentes conjuntos, especialmente cuando las observaciones de una distribución dada presentan datos muy dispersos.

## <font color=#56B4E9>5.1 Desviación media absoluta</font>
***


# $$DM = \frac 1n\sum_{i=1}^{n}|X_i-\bar{X}|$$


## <font color=#56B4E9>5.2 Varianza</font>
***

### Varianza

La varianza se construye a partir de las diferencias entre cada observación y la media de los datos, es decir, la desviación alrededor de la media. Al calcular la varianza, las desviaciones alrededor de la media son elevadas al cuadrado.

### Varianza de la población

# $$\sigma^2 = \frac 1n\sum_{i=1}^{n}(X_i-\mu)^2$$

### Varianza de la muestra

# $$S^2 = \frac 1{n-1}\sum_{i=1}^{n}(X_i-\bar{X})^2$$

## <font color=#56B4E9>5.3 Desviación estándar</font>
***

Una de las restricciones de la varianza es el hecho de que proporciona medidas cuadráticas de las unidades originales; la varianza de las medidas de longitud, por ejemplo, está en unidades de área. Por lo tanto, el hecho de que las unidades sean diferentes dificulta la comparación de la dispersión con las variables que la definen. Una forma de eliminar esta dificultad es considerar su raíz cuadrada.

### Desviación estándar de la población

# $$\sigma = \sqrt{\frac 1n\sum_{i=1}^{n}(X_i-\mu)^2} \Longrightarrow \sigma = \sqrt{\sigma^2}$$


### Desviación estándar de la muestra

# $$S = \sqrt{\frac 1{n-1}\sum_{i=1}^{n}(X_i-\bar{X})^2} \Longrightarrow S = \sqrt{S^2}$$